src/data/set/lattice.lean

```diff
@@ -744,6 +744,21 @@ begin
   simp [union_comm]
 end
 
+lemma union_Union_neq_eq_Union {ι} (f : ι → set α) (j : ι) :
+  (f j ∪ ⋃ (i : ι) (hi : i ≠ j), f i) = ⋃ i, f i :=
+begin
+  ext x,
+  simp only [exists_prop, mem_Union, mem_union_eq],
+  split,
+  { rintro (hj | ⟨i, hij, hi⟩),
+    { exact ⟨j, hj⟩ },
+    { exact ⟨i, hi⟩ } },
+  { rintro ⟨i, hi⟩,
+    by_cases h : i = j,
+    { exact or.inl (h ▸ hi) },
+    { exact or.inr ⟨i, h, hi⟩ } }
+end
```

Comment:
Is this true more generally for `supr`?

Reply:
I would imagine the `supr` version looking like 
```lean
example {β : Type*} [has_Sup β] [has_sup β] {ι} 
  (f : ι → β) (j : ι) : 
  (f j ⊔ ⨆ (i : ι) (hi : i ≠ j), f i) = ⨆ i, f i :=
begin
```
though I don't think this is provable in this state since `Sup` and `sup` are just arbitrary functions and does not necessarily agree

Reply:
I would have thought a complete lattice structure forces them to agree

Reply:
With complete lattice, the lemma comes out nicely
```lean
example {β : Type*} [complete_lattice β] {ι} 
  (f : ι → β) (j : ι) : (f j ⊔ ⨆ (i : ι) (hi : i ≠ j), f i) = ⨆ i, f i :=
begin
  apply le_antisymm,
  { exact sup_le (le_supr _ _) (supr_le (λ _, supr_le (λ _, le_supr _ _))) },
  { refine supr_le (λ i, _),
    by_cases h : i = j,
    { exact h ▸ le_sup_left },
    { exact le_sup_of_le_right (le_bsupr i h) } }
end```
I will change to this lemma later since I don't want to recompile the entirety of mathlib

Reply:
It probably makes sense to make a new PR containing that lemma, and then you can just merge master when you're ready to switch.

Reply:
After looking around, it seems that this lemma is just the symmetric version of `supr_split_single` so there is no need for a PR :p

Reply:
Here's a more atomic version of that proof:
```
lemma bsupr_sup_bsupr_compl {ι β : Type*} [complete_lattice β] (p : ι → Prop)
  (f : ι → β) :
    (⨆ i (hi : p i), f i) ⊔ (⨆ i (hi : ¬p i), f i) = ⨆ i, f i :=
begin
  rw ←supr_sup_eq,
  congr' with i,
  by_cases h : p i; simp [h],
end

lemma bsupr_eq {ι β : Type*} [complete_lattice β] (f : ι → β) (j : ι) :
  (⨆ (i : ι) (hi : i = j), f i) = f j :=
supr_singleton

lemma sup_bsupr_ne {ι β : Type*} [complete_lattice β] (f : ι → β) (j : ι) :
  (f j ⊔ ⨆ (i : ι) (hi : i ≠ j), f i) = ⨆ i, f i :=
begin
  rw ←bsupr_eq f j,
  exact bsupr_sup_bsupr_compl (= j) f
end
```

Reply:
Perhaps those lemmas need renaming or need aliases? IMO they're stated backwards, which makes them hard to find.

Reply:
It might be worth while to flip them and make them simp lemmas? I certainly think the lhs is more simple than the rhs in `supr_split_single`